File: extra-addons/odoocast-addons/tk_renewable_energy/models/renewable_energy_dashboard.py
# -*- coding: utf-8 -*-
# Copyright 2022-Today TechKhedut.
# Part of TechKhedut. See LICENSE file for full copyright and licensing details.
from odoo import models, fields, api


class RenewableEnergyDashboard(models.Model):
    _name = "renewable.energy.dashboard"
    _description = "Renewable Energy Dashboard"

    @api.model
    def get_renewable_energy_dashboard(self):
        total_renewable_energy_order = self.env['renewable.energy.order'].sudo().search_count([])
        order_site_inspection = self.env['renewable.energy.order'].sudo().search_count(
            [('order_stages', '=', 'site_inspection')])
        material_order_stages = self.env['renewable.energy.order'].sudo().search_count(
            [('order_stages', '=', 'material_order')])
        material_arrived_stages = self.env['renewable.energy.order'].sudo().search_count(
            [('order_stages', '=', 'material_arrived')])
        project_installation_stages = self.env['renewable.energy.order'].sudo().search_count(
            [('order_stages', '=', 'project_installation')])
        qa_check_stages = self.env['renewable.energy.order'].sudo().search_count([('order_stages', '=', 'qa_check')])
        complete_stages = self.env['renewable.energy.order'].sudo().search_count([('order_stages', '=', 'complete')])

        survey_team_task = self.env['project.task'].sudo().search_count([('team_type', '=', 'survey_team')])
        installation_team_task = self.env['project.task'].sudo().search_count([('team_type', '=', 'installation_team')])
        qa_team_task = self.env['project.task'].sudo().search_count([('team_type', '=', 'qa_team')])

        total_order_count_graph = [['Total Orders', 'In Site Inspection', 'In Material Order', 'In Material Arrived',
                                    'In Project Installation', 'In QA Check', 'Completed'],
                                   [total_renewable_energy_order, order_site_inspection, material_order_stages,
                                    material_arrived_stages, project_installation_stages, qa_check_stages,
                                    complete_stages]]
        project_team_task = [['Survey Team', 'Installation Team', 'QA Team'],
                             [survey_team_task, installation_team_task, qa_team_task]]

        total_ren_lead = self.env['ren.lead'].sudo().search_count([])
        ren_lead_in_discussion = self.env['ren.lead'].sudo().search_count([('status', '=', 'b_in_discussion')])
        ren_lead_won = self.env['ren.lead'].sudo().search_count([('status', '=', 'c_won')])
        ren_lead_lost = self.env['ren.lead'].sudo().search_count([('status', '=', 'd_lost')])

        total_lead_status = [['Total', 'In Discussion', 'Won', 'Lost'],
                             [total_ren_lead, ren_lead_in_discussion, ren_lead_won, ren_lead_lost]]

        lead_source_phone_calls = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'phone_calls')])
        lead_source_e_mail = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'e_mail')])
        lead_source_social_media = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'social_media')])
        lead_source_website = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'website')])
        lead_source_direct_mail = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'direct_mail')])
        lead_source_events = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'events')])
        lead_source_public_relations = self.env['ren.lead'].sudo().search_count(
            [('lead_source_type', '=', 'public_relations')])
        lead_source_content_marketing = self.env['ren.lead'].sudo().search_count(
            [('lead_source_type', '=', 'content_marketing')])
        lead_source_branding = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'branding')])
        lead_source_online_marketing = self.env['ren.lead'].sudo().search_count(
            [('lead_source_type', '=', 'online_marketing')])
        lead_source_referrals = self.env['ren.lead'].sudo().search_count([('lead_source_type', '=', 'referrals')])

        lead_source_status = [
            ['Phone Calls', 'E-mail', 'Social Media', 'Website', 'Direct Mail', 'Events', 'Public Relations',
             'Content Marketing', 'Branding', 'Online Marketing', 'Referrals'],
            [lead_source_phone_calls, lead_source_e_mail, lead_source_social_media,
             lead_source_website, lead_source_direct_mail, lead_source_events, lead_source_public_relations,
             lead_source_content_marketing, lead_source_branding, lead_source_online_marketing, lead_source_referrals]]

        lead_customer_residential = self.env['ren.lead'].sudo().search_count(
            [('customer_category', '=', 'residential')])
        lead_customer_institutional = self.env['ren.lead'].sudo().search_count(
            [('customer_category', '=', 'institutional')])
        lead_customer_industrial = self.env['ren.lead'].sudo().search_count([('customer_category', '=', 'industrial')])
        lead_customer_commerical = self.env['ren.lead'].sudo().search_count([('customer_category', '=', 'commerical')])
        lead_customer_government = self.env['ren.lead'].sudo().search_count([('customer_category', '=', 'government')])
        lead_customer_social_sector = self.env['ren.lead'].sudo().search_count(
            [('customer_category', '=', 'social_sector')])

        lead_customer_category = [
            ['Residential', 'Institutional', 'Industrial', 'Commerical', 'Government', 'Social Sector'],
            [lead_customer_residential, lead_customer_institutional, lead_customer_industrial,
             lead_customer_commerical, lead_customer_government, lead_customer_social_sector]]

        data = {
            'total_renewable_energy_order': total_renewable_energy_order,
            'order_site_inspection': order_site_inspection,
            'material_order_stages': material_order_stages,
            'material_arrived_stages': material_arrived_stages,
            'project_installation_stages': project_installation_stages,
            'qa_check_stages': qa_check_stages,
            'complete_stages': complete_stages,
            'survey_team_task': survey_team_task,
            'installation_team_task': installation_team_task,
            'qa_team_task': qa_team_task,
            'total_order_count_graph': total_order_count_graph,
            'project_team_task': project_team_task,
            'total_lead_status': total_lead_status,
            'lead_source_status': lead_source_status,
            'lead_customer_category': lead_customer_category,
        }
        return data

File: extra-addons/odoocast-addons/tk_renewable_energy/static/src/js/renewable_energy_dashboard.js
odoo.define('tk_renewable_energy.RenewableEnergyDashboard', function (require) {
    'use strict';
    const AbstractAction = require('web.AbstractAction');
    const ajax = require('web.ajax');
    const core = require('web.core');
    const rpc = require('web.rpc');
    const session = require('web.session')
    const web_client = require('web.web_client');
    const _t = core._t;
    const QWeb = core.qweb;
    const { loadBundle } = require("@web/core/assets");

    const ActionMenu = AbstractAction.extend({
        template: 'renewable_energy_dashboard_template',
        events: {
            'click .total-renewable-energy-order': 'view_total_renewable_energy_order',
            'click .order-site-inspection': 'view_order_site_inspection',
            'click .material-order-stages': 'view_material_order_stages',
            'click .material-arrived-stages': 'view_material_arrived_stages',
            'click .project-installation-stages': 'view_project_installation_stages',
            'click .qa-check-stages': 'view_qa_check_stages',
            'click .complete-stages': 'view_complete_stages',
            'click .survey-team-task': 'view_survey_team_task',
            'click .installation-team-task': 'view_installation_team_task',
            'click .qa-team-task': 'view_qa_team_task',
        },

        renderElement: function (ev) {
            const self = this;
            $.when(this._super())
                .then(function (ev) {
                    rpc.query({
                        model: "renewable.energy.dashboard",
                        method: "get_renewable_energy_dashboard",
                    }).then(function (result) {
                        $('#total_renewable_energy_order').empty().append(result['total_renewable_energy_order']);
                        $('#order_site_inspection').empty().append(result['order_site_inspection']);
                        $('#material_order_stages').empty().append(result['material_order_stages']);
                        $('#material_arrived_stages').empty().append(result['material_arrived_stages']);
                        $('#project_installation_stages').empty().append(result['project_installation_stages']);
                        $('#qa_check_stages').empty().append(result['qa_check_stages']);
                        $('#complete_stages').empty().append(result['complete_stages']);
                        $('#survey_team_task').empty().append(result['survey_team_task']);
                        $('#installation_team_task').empty().append(result['installation_team_task']);
                        $('#qa_team_task').empty().append(result['qa_team_task']);
                        self.renOrderStatusGraph(result['total_order_count_graph']);
                        self.getProjectTeamTask(result['project_team_task']);
                        self.getTotalLeadStatus(result['total_lead_status']);
                        self.getLeadSourceStatus(result['lead_source_status']);
                        self.getLeadCustomerCategory(result['lead_customer_category']);
                    });
                });
        },

        view_total_renewable_energy_order : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('REN Orders'),
                type: 'ir.actions.act_window',
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_order_site_inspection : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Site Inspections'),
                type: 'ir.actions.act_window',
                domain: [['order_stages', '=', 'site_inspection']],
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_material_order_stages : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Materials Order'),
                type: 'ir.actions.act_window',
                domain: [['order_stages', '=', 'material_order']],
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_material_arrived_stages : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Materials Arrived'),
                type: 'ir.actions.act_window',
                domain: [['order_stages', '=', 'material_arrived']],
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_project_installation_stages : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Project Installations'),
                type: 'ir.actions.act_window',
                domain: [['order_stages', '=', 'project_installation']],
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_qa_check_stages : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('QA Checks'),
                type: 'ir.actions.act_window',
                domain: [['order_stages', '=', 'qa_check']],
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_complete_stages : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Completes'),
                type: 'ir.actions.act_window',
                domain: [['order_stages', '=', 'complete']],
                res_model: 'renewable.energy.order',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'activity']],
                target: 'current'
            });
        },
        view_survey_team_task : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Survey Team Tasks'),
                type: 'ir.actions.act_window',
                domain: [['team_type', '=', 'survey_team']],
                res_model: 'project.task',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'pivot'], [false, 'graph'], [false, 'activity']],
                target: 'current'
            });
        },
        view_installation_team_task : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('Installation Team Tasks'),
                type: 'ir.actions.act_window',
                domain: [['team_type', '=', 'installation_team']],
                res_model: 'project.task',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'pivot'], [false, 'graph'], [false, 'activity']],
                target: 'current'
            });
        },
        view_qa_team_task : function (ev){
            ev.preventDefault();
            return this.do_action({
                name: _t('QA Team Tasks'),
                type: 'ir.actions.act_window',
                domain: [['team_type', '=', 'qa_team']],
                res_model: 'project.task',
                views: [[false, 'kanban'], [false, 'list'], [false, 'form'], [false, 'calendar'], [false, 'pivot'], [false, 'graph'], [false, 'activity']],
                target: 'current'
            });
        },

        get_action: function (ev, name, res_model){
            ev.preventDefault();
            return this.do_action({
                name: _t(name),
                type: 'ir.actions.act_window',
                res_model: res_model,
                views: [[false, 'kanban'],[false, 'tree'],[false, 'form']],
                target: 'current'
            });
        },

        apexGraph: function () {
            Apex.grid = {
                padding: {
                    right: 0,
                    left: 0,
                    top: 10,
                }
            }
            Apex.dataLabels = {
                enabled: false
            }
        },

        renOrderStatusGraph: function(data){
        const options = {
          series: [
            {
            name: 'Orders',
            data: data[1]
            }
          ],
        chart: {
          height: 400,
          type: 'bar',
        },
        colors: ['#f29e4c', '#f1c453', '#efea5a',  '#b9e769',  '#83e377', '#16db93', '#0db39e', '#048ba8', '#2c699a', '#54478c'],
        plotOptions: {
          bar: {
            columnWidth: '10%',
            distributed: true,
          }
        },
        dataLabels: {
          enabled: false
        },
        legend: {
          show: false
        },
        xaxis: {
            categories: data[0],
            labels: {
                style: {
                    colors: ['#f29e4c', '#f1c453', '#efea5a',  '#b9e769',  '#83e377', '#16db93', '#0db39e', '#048ba8', '#2c699a', '#54478c'],
                    fontSize: '12px'}
            }
        }
        };
        this.renderGraph("#ren_order_status", options);
        },

        getProjectTeamTask: function(data){
          const options = {
            series: data[1],
            chart: {
            width: 500,
            height: 500,
            type: 'polarArea',
          },
          yaxis: {
        labels: {
              formatter: function (val) {
                return val.toFixed(0);
              }
            },
        },
          labels: data[0],
          stroke: {
            colors: ['#fff']
          },
          fill: {
            opacity: 0.8
          },
          responsive: [{
            breakpoint: 480,
            options: {
                chart: {
                  width: 200
                },
                legend: {
                  position: 'bottom'
                }
            }
        }]
        };
        this.renderGraph("#project_team_task", options);
        },

        getTotalLeadStatus: function(data){
          const options = {
            series: data[1],
            chart: {
            width: 500,
            height: 500,
            type: 'polarArea',
          },
          yaxis: {
        labels: {
              formatter: function (val) {
                return val.toFixed(0);
              }
            },
        },
          labels: data[0],
          stroke: {
            colors: ['#fff']
          },
          fill: {
            opacity: 0.8
          },
          responsive: [{
            breakpoint: 480,
            options: {
                chart: {
                  width: 200
                },
                legend: {
                  position: 'bottom'
                }
            }
        }]
        };
        this.renderGraph("#total_lead_status", options);
        },

        getLeadSourceStatus: function(data){
          const options = {
            series: data[1],
            chart: {
            width: 500,
            height: 500,
            type: 'polarArea',
          },
          yaxis: {
        labels: {
              formatter: function (val) {
                return val.toFixed(0);
              }
            },
        },
          labels: data[0],
          stroke: {
            colors: ['#fff']
          },
          fill: {
            opacity: 0.8
          },
          responsive: [{
            breakpoint: 480,
            options: {
                chart: {
                  width: 200
                },
                legend: {
                  position: 'bottom'
                }
            }
        }]
        };
        this.renderGraph("#lead_source_status", options);
        },

        getLeadCustomerCategory: function(data){
          const options = {
            series: data[1],
            chart: {
            width: 500,
            height: 500,
            type: 'polarArea',
          },
          yaxis: {
        labels: {
              formatter: function (val) {
                return val.toFixed(0);
              }
            },
        },
          labels: data[0],
          stroke: {
            colors: ['#fff']
          },
          fill: {
            opacity: 0.8
          },
          responsive: [{
            breakpoint: 480,
            options: {
                chart: {
                  width: 200
                },
                legend: {
                  position: 'bottom'
                }
            }
        }]
        };
        this.renderGraph("#lead_customer_category", options);
        },

        renderGraph: function(render_id, options){
           $(render_id).empty();
           const graphData = new ApexCharts(document.querySelector(render_id), options);
           graphData.render();
        },

        willStart: function () {
            const self = this;
            self.drpdn_show = false;
            return Promise.all([loadBundle(this), this._super()]);
        },
    });
    core.action_registry.add('renewable_energy_dashboard', ActionMenu);
});
